### DISCUSSION TITLE
4D dark ride ride

### DISCUSSION THREAD

**Discussion by spacek531** [OpenRCT2/OpenRCT2#16928] [Feature requests]
A dark ride with rotation would be useful as a custom ride or official ride type. It would have rotation like 4D, but with the multidim rotation field representing a single 16-frame rotation, instead of two 8-frame rotations.

![image](https://user-images.githubusercontent.com/3454156/161650004-cd2774b1-8b56-4280-b574-c0ff9aed2c00.png)
(Example by @oli414)

To save on frames, the ride should be painted as parent and child: a chassis and a vehicle, to reuse as many frames as possible. This would necessitate a custom PaintType.

Some implementation ideas:
- When the vehicle is oriented such that the Vehicle is above the Chassis respective to the camera, paint the Chassis as a parent and the Vehicle as a child, then paint the peeps as a child.
- When the vehicle is oriented such that the Chassis is above the Vehicle, paint the Vehicle as a parent, paint the peeps as children, then paint the Chassis as a child.
- The above two bullet points allow the Vehicle and Chassis sprites to be rendered without occlusions with each other baked into the sprites. It does suggest that a shear line should be baked into the model such that no Chassis geometry extends above and no Vehicle geometry extends below. I think this will not be such a difficult target.
- If a Chassis is not needed, its sprites can be left blank. It will likely be easier to require the sprites for a Chassis in all cases and the rides that don't require the Chassis typically are limited in sprite groups.
- The PaintType could easily work with spinning the way it works with animation.
- Variations can be added to effect vehicles with different symmetries; 2-, 3-, 4-, 5-, and 6-way rotational symmetry ride vehicles exist in the forms of Rocking Tug, Disk'O, Maxi-Air, Giant Frisbee, several spinning coasters, and more.

**Comment by ocalhoun6** [comment]
Seems like the kind of thing that the Haunted Mansion Ride and/or Ghost Train _should_ have been, but the original RCT2 developers didn't add that functionality.

Making a brand new ride that does this is _one_ option ... but maybe another option would be to add a vehicle rotation capability to one of those two existing rides?

-----

> Vehicle sprite count would shake out like this per sprite group (not totaled):
> Flat sprites: 32
> Gentle slopes: 576
> Steep slopes: 640
> Diagonal slopes: 384
> Vertical slopes: 1344

Could cut down on this a lot by not allowing the ride to have steep slopes or vertical slopes, right? Doesn't seem like the sort of ride that should be going steep or vertical anyway ... much like the Haunted Mansion or Ghost Train rides mentioned before. Could you point out any real-world example that uses steep or vertical slopes?

Maybe even the diagonal slopes could be cut out, since neither of those two rides has access to diagonal sections either. (Though I do admit that having diagonal sections would be cool, and it would help distinguish this ride from the other two, so there's a good case for keeping the diagonal sections.)

**Reply by spacek531** [reply]
>Seems like the kind of thing that the Haunted Mansion Ride and/or Ghost Train should have been, but the original RCT2 developers didn't add that functionality.
> 
> Making a brand new ride that does this is one option ... but maybe another option would be to add a vehicle rotation capability to one of those two existing rides?

There's an intriguing reason this can't be done, and it has to do with OpenRCT2's RCT1 compatibility. Ghost train terrain doors use the exact same set of bits as 4D rotation, and, well, ghost train is the only ride type to have them (Haunted mansion uses the same ride type).

As far as "cutting down on this" there's not really any cutting down on them at all except by choosing not to use the various sprite groups.

**Comment by oli414** [comment]
Ideally one would be able to specify how many frames of animation it has. Just making a ride type with 16 animation frames, and a separate chassis makes this too purpose build. "4D dark ride" is not a great name for this btw

Do you have a practical example of what this might look like in the ParkObj?

**Reply by spacek531** [reply]
What do you mean "in the ParkObj?" It will either be a sprite atlas or image archive and a json config file. For the purposes of this kind of ride, any multiple of 4 will work for number of frames, but should also be a power of 2 lest it make countless unnecessary flat sprites.
